# Go API client for genome_nexus_internal_api

This page shows how to use HTTP requests to access the Genome Nexus API. There are more high level clients available in Python, R, JavaScript, TypeScript and various other languages as well as a command line client to annotate MAF and VCF. See https://docs.genomenexus.org/api.

Aside from programmatic clients there are web based tools to annotate variants, see https://docs.genomenexus.org/tools.

 We currently only provide long-term support for the '/annotation' endpoint. The other endpoints might change.

## Overview
This API client was generated by the [OpenAPI Generator](https://openapi-generator.tech) project.  By using the [OpenAPI-spec](https://www.openapis.org/) from a remote server, you can easily generate an API client.

- API version: 2.0
- Package version: 1.0.0
- Build package: org.openapitools.codegen.languages.GoClientCodegen

## Installation

Install the following dependencies:

```shell
go get github.com/stretchr/testify/assert
go get golang.org/x/oauth2
go get golang.org/x/net/context
```

Put the package under your project folder and add the following in import:

```golang
import genome_nexus_internal_api "github.com/GIT_USER_ID/GIT_REPO_ID"
```

To use a proxy, set the environment variable `HTTP_PROXY`:

```golang
os.Setenv("HTTP_PROXY", "http://proxy_name:proxy_port")
```

## Configuration of Server URL

Default configuration comes with `Servers` field that contains server objects as defined in the OpenAPI specification.

### Select Server Configuration

For using other server than the one defined on index 0 set context value `sw.ContextServerIndex` of type `int`.

```golang
ctx := context.WithValue(context.Background(), genome_nexus_internal_api.ContextServerIndex, 1)
```

### Templated Server URL

Templated server URL is formatted using default variables from configuration or from context value `sw.ContextServerVariables` of type `map[string]string`.

```golang
ctx := context.WithValue(context.Background(), genome_nexus_internal_api.ContextServerVariables, map[string]string{
	"basePath": "v2",
})
```

Note, enum values are always validated and all unused variables are silently ignored.

### URLs Configuration per Operation

Each operation can use different server URL defined using `OperationServers` map in the `Configuration`.
An operation is uniquely identified by `"{classname}Service.{nickname}"` string.
Similar rules for overriding default operation server index and variables applies by using `sw.ContextOperationServerIndices` and `sw.ContextOperationServerVariables` context maps.

```golang
ctx := context.WithValue(context.Background(), genome_nexus_internal_api.ContextOperationServerIndices, map[string]int{
	"{classname}Service.{nickname}": 2,
})
ctx = context.WithValue(context.Background(), genome_nexus_internal_api.ContextOperationServerVariables, map[string]map[string]string{
	"{classname}Service.{nickname}": {
		"port": "8443",
	},
})
```

## Documentation for API Endpoints

All URIs are relative to *http://www.genomenexus.org*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*AnnotationControllerApi* | [**FetchVariantAnnotationByGenomicLocationGET**](docs/AnnotationControllerApi.md#fetchvariantannotationbygenomiclocationget) | **Get** /annotation/genomic/{genomicLocation} | Retrieves VEP annotation for the provided genomic location
*AnnotationControllerApi* | [**FetchVariantAnnotationByGenomicLocationPOST**](docs/AnnotationControllerApi.md#fetchvariantannotationbygenomiclocationpost) | **Post** /annotation/genomic | Retrieves VEP annotation for the provided list of genomic locations
*AnnotationControllerApi* | [**FetchVariantAnnotationByIdGET**](docs/AnnotationControllerApi.md#fetchvariantannotationbyidget) | **Get** /annotation/dbsnp/{variantId} | Retrieves VEP annotation for the give dbSNP id
*AnnotationControllerApi* | [**FetchVariantAnnotationByIdPOST**](docs/AnnotationControllerApi.md#fetchvariantannotationbyidpost) | **Post** /annotation/dbsnp/ | Retrieves VEP annotation for the provided list of dbSNP ids
*AnnotationControllerApi* | [**FetchVariantAnnotationGET**](docs/AnnotationControllerApi.md#fetchvariantannotationget) | **Get** /annotation/{variant} | Retrieves VEP annotation for the provided variant
*AnnotationControllerApi* | [**FetchVariantAnnotationPOST**](docs/AnnotationControllerApi.md#fetchvariantannotationpost) | **Post** /annotation | Retrieves VEP annotation for the provided list of variants
*EnsemblControllerApi* | [**FetchCanonicalEnsemblGeneIdByEntrezGeneIdGET**](docs/EnsemblControllerApi.md#fetchcanonicalensemblgeneidbyentrezgeneidget) | **Get** /ensembl/canonical-gene/entrez/{entrezGeneId} | Retrieves Ensembl canonical gene id by Entrez Gene Id
*EnsemblControllerApi* | [**FetchCanonicalEnsemblGeneIdByEntrezGeneIdsPOST**](docs/EnsemblControllerApi.md#fetchcanonicalensemblgeneidbyentrezgeneidspost) | **Post** /ensembl/canonical-gene/entrez | Retrieves canonical Ensembl Gene ID by Entrez Gene Ids
*EnsemblControllerApi* | [**FetchCanonicalEnsemblGeneIdByHugoSymbolGET**](docs/EnsemblControllerApi.md#fetchcanonicalensemblgeneidbyhugosymbolget) | **Get** /ensembl/canonical-gene/hgnc/{hugoSymbol} | Retrieves Ensembl canonical gene id by Hugo Symbol
*EnsemblControllerApi* | [**FetchCanonicalEnsemblGeneIdByHugoSymbolsPOST**](docs/EnsemblControllerApi.md#fetchcanonicalensemblgeneidbyhugosymbolspost) | **Post** /ensembl/canonical-gene/hgnc | Retrieves canonical Ensembl Gene ID by Hugo Symbols
*EnsemblControllerApi* | [**FetchCanonicalEnsemblTranscriptByHugoSymbolGET**](docs/EnsemblControllerApi.md#fetchcanonicalensembltranscriptbyhugosymbolget) | **Get** /ensembl/canonical-transcript/hgnc/{hugoSymbol} | Retrieves Ensembl canonical transcript by Hugo Symbol
*EnsemblControllerApi* | [**FetchCanonicalEnsemblTranscriptsByHugoSymbolsPOST**](docs/EnsemblControllerApi.md#fetchcanonicalensembltranscriptsbyhugosymbolspost) | **Post** /ensembl/canonical-transcript/hgnc | Retrieves Ensembl canonical transcripts by Hugo Symbols
*EnsemblControllerApi* | [**FetchEnsemblTranscriptByTranscriptIdGET**](docs/EnsemblControllerApi.md#fetchensembltranscriptbytranscriptidget) | **Get** /ensembl/transcript/{transcriptId} | Retrieves the transcript by an Ensembl transcript ID
*EnsemblControllerApi* | [**FetchEnsemblTranscriptsByEnsemblFilterPOST**](docs/EnsemblControllerApi.md#fetchensembltranscriptsbyensemblfilterpost) | **Post** /ensembl/transcript | Retrieves Ensembl Transcripts by Ensembl transcript IDs, hugo Symbols, protein IDs, or gene IDs
*EnsemblControllerApi* | [**FetchEnsemblTranscriptsGET**](docs/EnsemblControllerApi.md#fetchensembltranscriptsget) | **Get** /ensembl/transcript | Retrieves Ensembl Transcripts by protein ID, and gene ID. Retrieves all transcripts in case no query parameter provided
*EnsemblControllerApi* | [**FetchGeneXrefsGET**](docs/EnsemblControllerApi.md#fetchgenexrefsget) | **Get** /ensembl/xrefs | Perform lookups of Ensembl identifiers and retrieve their external references in other databases
*InfoControllerApi* | [**FetchVersionGET**](docs/InfoControllerApi.md#fetchversionget) | **Get** /version | Retrieve Genome Nexus Version
*PdbControllerApi* | [**FetchPdbHeaderGET**](docs/PdbControllerApi.md#fetchpdbheaderget) | **Get** /pdb/header/{pdbId} | Retrieves PDB header info by a PDB id
*PdbControllerApi* | [**FetchPdbHeaderPOST**](docs/PdbControllerApi.md#fetchpdbheaderpost) | **Post** /pdb/header | Retrieves PDB header info by a PDB id
*PfamControllerApi* | [**FetchPfamDomainsByAccessionGET**](docs/PfamControllerApi.md#fetchpfamdomainsbyaccessionget) | **Get** /pfam/domain/{pfamAccession} | Retrieves a PFAM domain by a PFAM domain ID
*PfamControllerApi* | [**FetchPfamDomainsByPfamAccessionPOST**](docs/PfamControllerApi.md#fetchpfamdomainsbypfamaccessionpost) | **Post** /pfam/domain | Retrieves PFAM domains by PFAM domain accession IDs
*PtmControllerApi* | [**FetchPostTranslationalModificationsByPtmFilterPOST**](docs/PtmControllerApi.md#fetchposttranslationalmodificationsbyptmfilterpost) | **Post** /ptm/experimental | Retrieves PTM entries by Ensembl Transcript IDs
*PtmControllerApi* | [**FetchPostTranslationalModificationsGET**](docs/PtmControllerApi.md#fetchposttranslationalmodificationsget) | **Get** /ptm/experimental | Retrieves PTM entries by Ensembl Transcript ID


## Documentation For Models

 - [AggregateSourceInfo](docs/AggregateSourceInfo.md)
 - [AlleleCount](docs/AlleleCount.md)
 - [AlleleFrequency](docs/AlleleFrequency.md)
 - [AlleleNumber](docs/AlleleNumber.md)
 - [Alleles](docs/Alleles.md)
 - [ArticleAbstract](docs/ArticleAbstract.md)
 - [Citations](docs/Citations.md)
 - [ClinVar](docs/ClinVar.md)
 - [Clinvar](docs/Clinvar.md)
 - [ClinvarAnnotation](docs/ClinvarAnnotation.md)
 - [ColocatedVariant](docs/ColocatedVariant.md)
 - [Cosmic](docs/Cosmic.md)
 - [CountByTumorType](docs/CountByTumorType.md)
 - [Dbsnp](docs/Dbsnp.md)
 - [Drug](docs/Drug.md)
 - [EnsemblFilter](docs/EnsemblFilter.md)
 - [EnsemblGene](docs/EnsemblGene.md)
 - [EnsemblTranscript](docs/EnsemblTranscript.md)
 - [Exon](docs/Exon.md)
 - [Gene](docs/Gene.md)
 - [GeneXref](docs/GeneXref.md)
 - [GeneralPopulationStats](docs/GeneralPopulationStats.md)
 - [GenomeNexusInfo](docs/GenomeNexusInfo.md)
 - [GenomicLocation](docs/GenomicLocation.md)
 - [Gnomad](docs/Gnomad.md)
 - [Hg19](docs/Hg19.md)
 - [Hg38](docs/Hg38.md)
 - [Hgvs](docs/Hgvs.md)
 - [Homozygotes](docs/Homozygotes.md)
 - [Hotspot](docs/Hotspot.md)
 - [HotspotAnnotation](docs/HotspotAnnotation.md)
 - [HrdScore](docs/HrdScore.md)
 - [Implication](docs/Implication.md)
 - [IndicatorQueryResp](docs/IndicatorQueryResp.md)
 - [IndicatorQueryTreatment](docs/IndicatorQueryTreatment.md)
 - [IntegerRange](docs/IntegerRange.md)
 - [IntergenicConsequences](docs/IntergenicConsequences.md)
 - [MainType](docs/MainType.md)
 - [MutationAssessor](docs/MutationAssessor.md)
 - [MutationAssessorAnnotation](docs/MutationAssessorAnnotation.md)
 - [MutationEffectResp](docs/MutationEffectResp.md)
 - [Mutdb](docs/Mutdb.md)
 - [MyVariantInfo](docs/MyVariantInfo.md)
 - [MyVariantInfoAnnotation](docs/MyVariantInfoAnnotation.md)
 - [NucleotideContext](docs/NucleotideContext.md)
 - [NucleotideContextAnnotation](docs/NucleotideContextAnnotation.md)
 - [OncokbAnnotation](docs/OncokbAnnotation.md)
 - [PdbHeader](docs/PdbHeader.md)
 - [PfamDomain](docs/PfamDomain.md)
 - [PfamDomainRange](docs/PfamDomainRange.md)
 - [PostTranslationalModification](docs/PostTranslationalModification.md)
 - [PtmAnnotation](docs/PtmAnnotation.md)
 - [PtmFilter](docs/PtmFilter.md)
 - [Query](docs/Query.md)
 - [Rcv](docs/Rcv.md)
 - [SignalAnnotation](docs/SignalAnnotation.md)
 - [SignalMutation](docs/SignalMutation.md)
 - [SignalPopulationStats](docs/SignalPopulationStats.md)
 - [Snpeff](docs/Snpeff.md)
 - [SourceVersionInfo](docs/SourceVersionInfo.md)
 - [StatsByTumorType](docs/StatsByTumorType.md)
 - [TranscriptConsequence](docs/TranscriptConsequence.md)
 - [TranscriptConsequenceSummary](docs/TranscriptConsequenceSummary.md)
 - [TumorType](docs/TumorType.md)
 - [UntranslatedRegion](docs/UntranslatedRegion.md)
 - [VEPInfo](docs/VEPInfo.md)
 - [VariantAnnotation](docs/VariantAnnotation.md)
 - [VariantAnnotationSummary](docs/VariantAnnotationSummary.md)
 - [Vcf](docs/Vcf.md)
 - [Version](docs/Version.md)
 - [Vues](docs/Vues.md)


## Documentation For Authorization

 Endpoints do not require authorization.


## Documentation for Utility Methods

Due to the fact that model structure members are all pointers, this package contains
a number of utility functions to easily obtain pointers to values of basic types.
Each of these functions takes a value of the given basic type and returns a pointer to it:

* `PtrBool`
* `PtrInt`
* `PtrInt32`
* `PtrInt64`
* `PtrFloat`
* `PtrFloat32`
* `PtrFloat64`
* `PtrString`
* `PtrTime`

## Author



